vsphere/host_network_policy_structure.go

```diff
@@ -0,0 +1,273 @@
+// HostNicTeamingPolicy.
+func expandHostNicTeamingPolicy(d *schema.ResourceData) *types.HostNicTeamingPolicy {
+	obj := &types.HostNicTeamingPolicy{
+		Policy: d.Get("teaming_policy").(string),
+	}
+	if v, ok := d.GetOkExists("failback"); ok {
+		obj.RollingOrder = &([]bool{!v.(bool)}[0])
+	}
+	if v, ok := d.GetOkExists("notify_switches"); ok {
+		obj.NotifySwitches = &([]bool{v.(bool)}[0])
+	}
+	obj.FailureCriteria = expandHostNicFailureCriteria(d)
+	obj.NicOrder = expandHostNicOrderPolicy(d)
+
+	// These fields are deprecated and are set only to make things work. They are
+	// not exposed to Terraform.
+	obj.ReversePolicy = &([]bool{true}[0])
+
+	return obj
+}
+
+// flattenHostNicTeamingPolicy reads various fields from a HostNicTeamingPolicy
+// into the passed in ResourceData.
+func flattenHostNicTeamingPolicy(d *schema.ResourceData, obj *types.HostNicTeamingPolicy) error {
+	if obj.RollingOrder != nil {
+		d.Set("failback", !*obj.RollingOrder)
+	}
+	if obj.NotifySwitches != nil {
+		d.Set("notify_switches", *obj.NotifySwitches)
+	}
+	d.Set("teaming_policy", obj.Policy)
+	if err := flattenHostNicFailureCriteria(d, obj.FailureCriteria); err != nil {
+		return err
+	}
+	if err := flattenHostNicOrderPolicy(d, obj.NicOrder); err != nil {
+		return err
+	}
+	return nil
+}
+
+// expandHostNetworkSecurityPolicy reads certain ResourceData keys and returns
+// a HostNetworkSecurityPolicy.
+func expandHostNetworkSecurityPolicy(d *schema.ResourceData) *types.HostNetworkSecurityPolicy {
+	obj := &types.HostNetworkSecurityPolicy{}
+	if v, ok := d.GetOkExists("allow_promiscuous"); ok {
+		obj.AllowPromiscuous = &([]bool{v.(bool)}[0])
+	}
+	if v, ok := d.GetOkExists("forged_transmits"); ok {
+		obj.ForgedTransmits = &([]bool{v.(bool)}[0])
+	}
+	if v, ok := d.GetOkExists("mac_changes"); ok {
+		obj.MacChanges = &([]bool{v.(bool)}[0])
+	}
+	return obj
+}
+
+// flattenHostNetworkSecurityPolicy reads various fields from a
+// HostNetworkSecurityPolicy into the passed in ResourceData.
+func flattenHostNetworkSecurityPolicy(d *schema.ResourceData, obj *types.HostNetworkSecurityPolicy) error {
+	if obj.AllowPromiscuous != nil {
+		d.Set("allow_promiscuous", *obj.AllowPromiscuous)
+	}
+	if obj.ForgedTransmits != nil {
+		d.Set("forged_transmits", *obj.ForgedTransmits)
+	}
+	if obj.MacChanges != nil {
+		d.Set("mac_changes", *obj.MacChanges)
+	}
+	return nil
+}
+
+// expandHostNetworkTrafficShapingPolicy reads certain ResourceData keys and
+// returns a HostNetworkTrafficShapingPolicy.
+func expandHostNetworkTrafficShapingPolicy(d *schema.ResourceData) *types.HostNetworkTrafficShapingPolicy {
+	obj := &types.HostNetworkTrafficShapingPolicy{
+		AverageBandwidth: int64(d.Get("shaping_average_bandwidth").(int)),
+		BurstSize:        int64(d.Get("shaping_burst_size").(int)),
+		PeakBandwidth:    int64(d.Get("shaping_peak_bandwidth").(int)),
+	}
+	if v, ok := d.GetOkExists("shaping_enabled"); ok {
```

Comment:
Hey @tombuildsstuff, I have been trying to avoid unnecessary nesting whenever possible. I have lost too many dev hours on trying to work around the issues of dealing with nested config stuff in TF - the inheritance model of this resource in conjunction with `vsphere_host_port_group` also makes things even more complicated. So going forward I am only reserving using lists and sets for things that need it because they need to deal with multiple values and that's it. A virtual switch or port group can only have one policy, of which traffic shaping options are a part of, so this does not need to be nested here.